### ISSUE TITLE
Minor memory leak without sqlite3_shutdown

### ISSUE THREAD

**Issue by gooderist** [utelle/SQLite3MultipleCiphers#29]
Hello, I was seeing some memory leaks on app exit and traced it to sqlite3mc_amalgamation.c(23761). Specifically, three blocks of size 8, 48, and 208 bytes. One of these being `wsdAutoext.aExt` which is freed in `sqlite3_reset_auto_extension` but only if I call `sqlite3_shutdown`. Not sure if I'm just doing something wrong, but the call to shutdown seems to be necessary for me (building the default sqlite3mc_amalgamation.c on VS2019).

Not that it's a significant amount of memory or anything, but the "Detected memory leaks!" warning might make someone think twice about using the library. Anyway, it might be nice to mention this somewhere in the docs for others who might run into the same leaks.

**Comment by utelle** [comment]
If you want to be on the safe side, you'll have to call `sqlite3_shutdown`. This is the only way to guarantee that **all** resources are freed.

Interestingly, I use VS2019 myself, but I didn't experience memory leaks, even without calling `sqlite3_shutdown`. For example, the shell application coming with SQLite does not call `sqlite3_shutdown`. Nevertheless VS2019 does not report memory leaks, if I use this shell together with _SQLite3 Multiple Ciphers_.

How do you compile `sqlite3mc_amalgamation.c`? Could you please show the full VS2019 command line which is used? TIA.

**Comment by gooderist** [comment]
Thanks for the confirmation! The official sqlite3 documentation is a little vague on when sqlite3_shutdown is necessary. But I will assume it is in this case.

Here is a minimal sample app that illustrates the leaking.
```c
// main.c
// --------------------------------------------------------------
// Illustration of memory leak with sqlite3mc on app exit
// when not calling sqlite3_shutdown().
//
// Compile with MSVC from the commandline with
// cl /MTd main.c
//
// Run with the debugger to see the memory leak output
// devenv main.exe
//
// --------------------------------------------------------------

// Enable memory leak detection
// Ref: https://docs.microsoft.com/en-us/visualstudio/debugger/finding-memory-leaks-using-the-crt-library?view=vs-2019
#define _CRTDBG_MAP_ALLOC
#include <stdlib.h>
#include <crtdbg.h>

// Simple unity build for demonstration purposes
#include "sqlite3mc_amalgamation.c"
#include "sqlite3mc_amalgamation.h"

int main(int argc, char *argv[])
{
    // Enable memory leak reporting (will show up in MSVC "Output" window)
    _CrtSetDbgFlag ( _CRTDBG_ALLOC_MEM_DF | _CRTDBG_LEAK_CHECK_DF );

    sqlite3* db;

    sqlite3_open(":memory:", &db);
    sqlite3_close(db);

    // Memory is leaked at app exit unless shutdown is explicitly called
    // sqlite3_shutdown();

    return 0;
}
```

**Comment by utelle** [comment]
> The official sqlite3 documentation is a little vague on when sqlite3_shutdown is necessary.

Yes, indeed. In fact, it depends on how the SQLite library was built, especially whether the standard SQLite amalgamation is used with default settings or additional features were added and/or activated (as is the case for _SQLite3 Multiple Ciphers_). In the default case `sqlite3_shutdown` does not need to free any memory, and therefore it is not necessary to call the function at all in that case.

> But I will assume it is in this case.

Yes. _SQLite3 Multiple Ciphers_ makes use of the function `sqlite3_auto_extension` to register extensions for automatic loading for each new database connection. And therefore it is necessary to call `sqlite3_reset_auto_extension` (which is done implicitly by `sqlite3_shutdown`) to clean up memory before exiting the application - this is true for the standard SQLite3 code, too. This could be avoided only at the price of patching the SQLite source code even more.

Additionally, _SQLite3 Multiple Ciphers_ registers at least one dynamically allocated VFS instance, which is freed only if `sqlite3_shutdown` is called or if `sqlite3mc_vfs_shutdown` (or `sqlite3mc_shutdown`) is called explicitly. At first I had implemented only a single static instance of the new VFS shim, but that was too restrictive, in case a developer wanted to combine encryption with a non-default VFS.

So, the conclusion is that the simplest way to avoid memory leaks is to call `sqlite3_shutdown` before exiting the application.

Thanks for bringing the issue to my attention. I will add a note to the documentation.
